File: Binary_Tree_Inorder_Traversal.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Inorder in vector which contains node values.
     */
public:
    vector<int> inorderTraversal(TreeNode *root) {
        // write your code here
        // stack<TreeNode*> vn;
        // vector<int> result;
        // TreeNode * cur = root;
        // while(!vn.empty() || cur){

        //     if(cur){
        //         vn.emplace(cur);
        //         cur = cur->left;
        //     }else{
        //         cur = vn.top();
        //         vn.pop();
        //         result.emplace_back(cur->val);
        //         cur = cur->right;
        //     }
        // }
        // return result;
        vector<int> result;
        TreeNode* cur = root;
        while(cur){
            if(!cur->left){
                result.emplace_back(cur->val);
                cur = cur->right;
            }else{
                TreeNode* node = cur->left;
                while(node->right && node->right != cur){
                    node = node->right;
                }
                if(!node->right){
                    node->right = cur;
                    cur = cur->left;
                }else{
                    result.emplace_back(cur->val);
                    node->right = nullptr;
                    cur = cur->right;
                }
            }
        }
        return result;
    }
};

File: Binary Tree Inorder Traversal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param root: The root of binary tree.
    @return: Inorder in ArrayList which contains node values.
    """
    def inorderTraversal(self, root):
        # write your code here
        # def doTra(node, ret):
        #     if not node:
        #         return
        #     else:
        #         doTra(node.left, ret)
        #         ret.append(node.val)
        #         doTra(node.right, ret)
        
        # ret = []
        # doTra(root, ret)
        # return ret
        
        # ret = []
        # stack = []
        # node = root
        # while node:
        #     stack.append(node)
        #     node = node.left
        # while stack:
        #     node = stack.pop()
        #     ret.append(node.val)
        #     if node.right:
        #         node = node.right
        #         while node:
        #             stack.append(node)
        #             node = node.left
        # return ret
        if not root:
            return []
        stack, ret = [], []
        cur = root
        while stack or cur:
            while cur:
                stack.append(cur)
                cur = cur.left
            cur = stack.pop()
            ret.append(cur.val)
            cur = cur.right
        return ret